eliot/_action.py

```diff
@@ -415,6 +424,286 @@ def __exit__(self, type, exception, traceback):
+                message, action, message.task_level, action.task_level))
+
+
+class WrongActionType(Exception):
+    """
+    Tried to end a message with a different action_type to the beginning.
+    """
+
+    def __init__(self, action, message):
+        error_msg = 'Tried to end {} with {}. Expected action_type = {}, got {}'
+        Exception.__init__(
+            self, error_msg.format(
+                action, message, action.action_type,
+                message.contents.get(ACTION_TYPE_FIELD, '<undefined>')))
+
+
+class InvalidStatus(Exception):
+    """
+    Tried to end a message with an invalid status.
+    """
+
+    def __init__(self, action, message):
+        error_msg = 'Tried to end {} with {}. Expected status {} or {}, got {}'
+        Exception.__init__(
+            self, error_msg.format(
+                action, message, SUCCEEDED_STATUS, FAILED_STATUS,
+                message.contents.get(ACTION_STATUS_FIELD, '<undefined>')))
+
+
+class DuplicateChild(Exception):
+    """
+    Tried to add a child to an action that already had a child at that task
+    level.
+    """
+
+    def __init__(self, action, message):
+        Exception.__init__(
+            self, 'Tried to add {} to {}, but already had child at {}'.format(
+                message, action, message.task_level))
+
+
+class InvalidStartMessage(Exception):
+    """
+    Tried to start an action with an invalid message.
+    """
+
+    def __init__(self, message, reason):
+        Exception.__init__(
+            self, 'Invalid start message {}: {}'.format(message, reason))
+
+    @classmethod
+    def wrong_status(cls, message):
+        return cls(message, 'must have status "STARTED"')
+
+    @classmethod
+    def wrong_task_level(cls, message):
+        return cls(message, 'first message must have task level ending in 1')
+
+
+class WrittenAction(PClass):
+    """
+    An Action that has been logged.
+
+    @ivar action_type: The type of the action,
+        e.g. C{"yourapp:subsystem:dosomething"}
+
+    @ivar status: One of C{STARTED_STATUS}, C{SUCCEEDED_STATUS}, or
+        C{FAILED_STATUS}.
+
+    @ivar task_uuid: The UUID of the task to which this action belongs.
+
+    @ivar TaskLevel task_level: The level of the task in which the action occurs.
+
+    @ivar start_time: The Unix timestamp of when the action started.
+
+    @ivar end_time: The Unix timestamp of when the action ended, or C{None} if
+        there has been no end message.
+
+    @ivar exception: If the action failed, the name of the exception that was
+        raised to cause it to fail. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+
+    @ivar reason: The reason the action failed. If the action succeeded, or
+        hasn't finished yet, then C{None}.
+
+    @ivar _children: A L{pvector} of L{WrittenAction} and L{WrittenMessage}
+        that make up this action.
+    """
+
+    # Broad philosophy:
```

Comment:
I've moved the general sense of it to the docstring. The more aspirational parts have been deleted.